hail/src/main/scala/is/hail/expr/ir/PValue.scala

```diff
@@ -1,27 +1,92 @@
 package is.hail.expr.ir
 
-import is.hail.asm4s.{Code, TypeInfo}
-import is.hail.expr.types.physical.{PType, PVoid}
+import is.hail.annotations.{Region, UnsafeUtils}
+import is.hail.asm4s._
+import is.hail.expr.types.physical.{PCanonicalArray, PCanonicalDict, PCanonicalSet, PContainer, PType, PVoid}
 
-abstract class PSettable {
-  def load(): PValue
+abstract class PSettable[PV <: PValue] {
+  def load(): PV
 
-  def store(v: PValue): Code[Unit]
+  def store(v: PV): Code[Unit]
 
-  def :=(v: PValue): Code[Unit] = store(v)
+  def :=(v: PV): Code[Unit] = store(v)
+
+  def storeAny(v: PValue): Code[Unit] = store(v.asInstanceOf[PV])
 }
 
 object PValue {
+  def apply(pt: PType, code: Code[_]): PValue = pt match {
+    case pt: PCanonicalArray =>
+      new PCanonicalIndexableValue(pt, coerce[Long](code))
+    case pt: PCanonicalSet =>
+      new PCanonicalIndexableValue(pt, coerce[Long](code))
+    case pt: PCanonicalDict =>
+      new PCanonicalIndexableValue(pt, coerce[Long](code))
+
+    case _ =>
```

Comment:
No base case; we should break when used on pt that is not PPrimitive

Reply:
That's not possible.  I don't support any of the other ptypes here yet, e.g. PStruct/Tuple.

Reply:
In that case would the intent of the function be clearer if you matched on PPrimitive directly? What are the disadvantages of doing so?

Reply:
I don't think I understand what you're asking for.

This (temporary) PValue constructor constructs the relevant PValue types from a ptype and a code. I currently have 2 pvalues: indexable and primitive.  Primitive handles all other ptypes, which can be represented by a primitive, that is, a Code[T].  That's why it is the default case.

So I don't what I'd do if I switched off PPrimitive.

Reply:
This was my suggestion -- I basically want a noisy match error or assertion failure when we add a new ptype that should be indexable (PFixedLengthArray, for instance) but fail to include it here.

Reply:
Sure, although this won't exist in the final code.  This constructor can't work in the long term because a PValue may not be a single Code[T].  It is a bridge from the existing world where everything is represented by a scalar to the new world being built up.

Reply:
yeah, but PFixedLengthArray can exist with a single Code[T] PValue. I do want to try to implement a new ptype soon (one with the same stack representation as canonical ptypes) to try the new infrastructure out.

Reply:
I'm trying to be extra attentive to anywhere we match on ptypes to make sure we try to make these places fail loudly when we extend the current system.

Reply:
And what's that world look like?  Like this:

```
-        EmitTriplet(setup, xmv, PValue(pt,
-          Region.loadIRIntermediate(x.pType)(pArray.elementOffset(xa.load().tcode[Long], len, xi))))
+        EmitTriplet(setup, xmv, xa.load().loadElement(len, xi))
```

Reply:
> I'm trying to be extra attentive to anywhere we match on ptypes

Sure, I hear that.  I think basically you never want to switch on ptypes.

I'm not sure what the suggestion is here, tho.

Reply:
This sounds like something that can be addressed in a future PR.